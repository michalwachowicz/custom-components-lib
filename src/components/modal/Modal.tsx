import * as styles from "./Modal.module.scss";

interface ModalProps {
  open?: boolean;
  onClose?: () => void;
  children?: React.ReactNode | string;
}

export default function Modal({ open, onClose, children }: ModalProps) {
  if (!open) return null;

  return (
    <div className={styles["modal-container"]}>
      <div
        data-testid='modal-backdrop'
        className={styles["modal-backdrop"]}
        onClick={onClose}
      />
      <div className={styles["modal"]} role='dialog' aria-modal>
        {children}
      </div>
    </div>
  );
}
