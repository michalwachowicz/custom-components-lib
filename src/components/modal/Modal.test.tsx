import { act } from "react";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import Modal from "./Modal";

describe("<Modal />", () => {
  it("displays the modal when open is true", () => {
    render(<Modal open />);

    expect(screen.getByRole("dialog")).toBeInTheDocument();
    expect(screen.getByTestId("modal-backdrop")).toBeInTheDocument();
  });

  it("does not display the modal when open is false", () => {
    render(<Modal>Test</Modal>);

    expect(screen.queryByRole("dialog")).not.toBeInTheDocument();
    expect(screen.queryByTestId("modal-backdrop")).not.toBeInTheDocument();
    expect(screen.queryByText("Test")).not.toBeInTheDocument();
  });

  it("shows the children when passed", () => {
    render(<Modal open>Test</Modal>);
    expect(screen.getByText("Test")).toBeInTheDocument();
  });

  it("closes the modal when backdrop is clicked", async () => {
    const handleClose = jest.fn();
    const user = userEvent.setup();

    render(<Modal open onClose={handleClose} />);

    await act(async () => {
      await user.click(screen.getByRole("dialog"));
    });

    expect(handleClose).toHaveBeenCalledTimes(0);

    await act(async () => {
      await user.click(screen.getByTestId("modal-backdrop"));
    });

    expect(handleClose).toHaveBeenCalledTimes(1);
  });

  describe("snapshots", () => {
    const children = ["Test", undefined];
    const onClose = [undefined, jest.fn()];
    const open = [true, false];

    const testCases = open.flatMap((o) =>
      onClose.flatMap((c) =>
        children.map((child) => ({
          open: o,
          onClose: c,
          children: child,
        }))
      )
    );

    test.each(testCases)(
      "renders correctly with open=%s, onClose=%s, children=%s",
      ({ open, onClose, children }) => {
        const { asFragment } = render(
          <Modal open={open} onClose={onClose}>
            {children}
          </Modal>
        );

        expect(asFragment()).toMatchSnapshot();
      }
    );
  });
});
