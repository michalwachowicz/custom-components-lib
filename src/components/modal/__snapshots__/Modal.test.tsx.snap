// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Modal /> snapshots renders correctly with open={ open: false, onClose: [Function], children: 'Test' }, onClose=%s, children=%s 1`] = `<DocumentFragment />`;

exports[`<Modal /> snapshots renders correctly with open={ open: false, onClose: [Function], children: undefined }, onClose=%s, children=%s 1`] = `<DocumentFragment />`;

exports[`<Modal /> snapshots renders correctly with open={ open: false, onClose: undefined, children: 'Test' }, onClose=%s, children=%s 1`] = `<DocumentFragment />`;

exports[`<Modal /> snapshots renders correctly with open={ open: false, onClose: undefined, children: undefined }, onClose=%s, children=%s 1`] = `<DocumentFragment />`;

exports[`<Modal /> snapshots renders correctly with open={ open: true, onClose: [Function], children: 'Test' }, onClose=%s, children=%s 1`] = `
<DocumentFragment>
  <div
    class="modal-container"
  >
    <div
      class="modal-backdrop"
      data-testid="modal-backdrop"
    />
    <div
      aria-modal="true"
      class="modal"
      role="dialog"
    >
      Test
    </div>
  </div>
</DocumentFragment>
`;

exports[`<Modal /> snapshots renders correctly with open={ open: true, onClose: [Function], children: undefined }, onClose=%s, children=%s 1`] = `
<DocumentFragment>
  <div
    class="modal-container"
  >
    <div
      class="modal-backdrop"
      data-testid="modal-backdrop"
    />
    <div
      aria-modal="true"
      class="modal"
      role="dialog"
    />
  </div>
</DocumentFragment>
`;

exports[`<Modal /> snapshots renders correctly with open={ open: true, onClose: undefined, children: 'Test' }, onClose=%s, children=%s 1`] = `
<DocumentFragment>
  <div
    class="modal-container"
  >
    <div
      class="modal-backdrop"
      data-testid="modal-backdrop"
    />
    <div
      aria-modal="true"
      class="modal"
      role="dialog"
    >
      Test
    </div>
  </div>
</DocumentFragment>
`;

exports[`<Modal /> snapshots renders correctly with open={ open: true, onClose: undefined, children: undefined }, onClose=%s, children=%s 1`] = `
<DocumentFragment>
  <div
    class="modal-container"
  >
    <div
      class="modal-backdrop"
      data-testid="modal-backdrop"
    />
    <div
      aria-modal="true"
      class="modal"
      role="dialog"
    />
  </div>
</DocumentFragment>
`;
